middle_end/flambda2/tests/algorithms/patricia_tree_tests.ml

```diff
@@ -0,0 +1,915 @@
+
+  let min_binding_vs_opt m =
+    Option.equal equal_bindings (Map.min_binding_opt m)
+      (option_of_not_found Map.min_binding m)
+
+  let max_binding_vs_opt m =
+    Option.equal equal_bindings (Map.max_binding_opt m)
+      (option_of_not_found Map.max_binding m)
+
+  let min_binding_opt m =
+    match Map.min_binding_opt m with
+    | None -> Map.is_empty m
+    | Some (k, v) ->
+      Map.find_opt k m =? Some v && Map.for_all (fun k' _ -> k' >= k) m
+
+  let max_binding_opt m =
+    match Map.max_binding_opt m with
+    | None -> Map.is_empty m
+    | Some (k, v) ->
+      Map.find_opt k m =? Some v && Map.for_all (fun k' _ -> k' <= k) m
+
+  let choose_vs_opt m =
+    Option.equal equal_bindings (Map.choose_opt m)
+      (option_of_not_found Map.choose m)
+
+  let choose_opt m =
+    match Map.choose_opt m with
+    | None -> Map.is_empty m
+    | Some (k, v) -> Map.find_opt k m =? Some v
+
+  let split_valid k m =
+    let m_lo, _, m_hi = Map.split k m in
+    Map.valid m_lo && Map.valid m_hi
+
+  let split k m =
+    let m_lo, binding, m_hi = Map.split k m in
+    binding =? Map.find_opt k m
+    && Map.equal V.equal m_lo (Map.filter (fun k' _ -> k' < k) m)
+    && Map.equal V.equal m_hi (Map.filter (fun k' _ -> k' > k) m)
+
+  let map_valid f m = Map.valid (Map.map f m)
+
+  let map f m =
+    List.equal equal_bindings
+      (Map.map f m |> Map.bindings)
+      (List.map (fun (k, v) -> k, f v) (m |> Map.bindings))
+
+  let mapi_valid f m = Map.valid (Map.mapi f m)
+
+  let mapi f m =
+    List.equal equal_bindings
+      (Map.mapi f m |> Map.bindings)
+      (List.map (fun (k, v) -> k, f k v) (m |> Map.bindings))
+
+  let to_seq_vs_bindings m =
+    List.equal equal_bindings
+      (m |> Map.to_seq |> List.of_seq |> List.sort compare_bindings)
+      (m |> Map.bindings)
+
+  let of_list_valid l = Map.valid (Map.of_list l)
+
+  module Equality_on_bindings = struct
+    let sort_by_key l = List.sort (fun (k1, _) (k2, _) -> Int.compare k1 k2) l
+
+    let sort_and_group_by_key l =
+      let rec groups l =
+        match l with
+        | [] -> []
+        | (k, v) :: l ->
+          let rec group l =
+            match l with
+            | [] -> [], []
+            | (k', v') :: l when k = k' ->
+              let g, l = group l in
+              v' :: g, l
+            | l -> [], l
+          in
+          let g, l = group l in
+          (k, v :: g) :: groups l
+      in
+      groups (sort_by_key l)
+
+    let same_bindings_up_to_duplicate_keys l1 l2 =
+      let l1 = l1 |> sort_and_group_by_key in
+      let l2 = l2 |> sort_and_group_by_key in
+      let rec check l1 l2 =
+        match l1, l2 with
+        | [], [] -> true
+        | [], _ | _, [] -> false
+        | (k1, vs1) :: l1, (k2, vs2) :: l2 ->
+          Int.equal k1 k2
+          && List.exists (fun v1 -> List.exists (V.equal v1) vs2) vs1
+          && check l1 l2
+      in
+      check l1 l2
+  end
+
+  open Equality_on_bindings
```

Comment:
Partly because otherwise the helper functions look like more propositions to test. (Though I'm much less convinced now that keeping them separate from the calls to `check` makes sense.)